docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+example of such a function is =mapcar= which takes a function and a list
+and applies the function to every item of the list.
+
+#+BEGIN_SRC elisp
+(defun app (fn)
+  "Apply FN to the list (1 2 3 4)"
+  (declare (elsa ((function (number) number)) (list number)))
+  (mapcar fn (list 1 2 3 4)))
+
+(app (lambda (x) (* x x)))
+#+END_SRC
+
+The =app= function requires that we pass in a function which processes a
+number into a number and returns a list of numbers.
+
+** Generic types
+
+Generic types are types where some of the type arguments are variable.
+Both basic and composite types can be turned into generic types.
+
+*** Motivation
+
+An example of a generic function is =identity=.  This function takes
+anything in and anything out.  We could therefore give it a type
+annotation =(elsa (mixed) mixed)=.
+
+However, we can do better!  We know that whatever was passed in will
+be returned and so the type actually must be the same.  The =(elsa
+(mixed) mixed)= signature allows us to pass in an =int= and it can return
+back a =string= no problem and so it would not catch a huge number of
+possible errors.
+
+What we want to express here is "X comes in, X comes out".
+
+*** Syntax
+
+The syntax for generic types is "generic type name" + =*= suffix.  Any
+string can be used for the generic type name, but customarily
+single-letter names are used.
+
+For the above mentioned identity function we therefore write the type
+as =(elsa (a*) a*)= where =a*= stands for a generic type =a=.
+
+A function such as =car= can be typed as follows:
+
+#+BEGIN_SRC elisp
+(elsa ((cons a* b*)) a*)
+#+END_SRC
+
+It takes a cons with =a= in the =car= and =b= in the =cdr= and return the =car=
+which is of type =a= , whatever that happens to be.
+
+** Optional types
+
+If a function can take optional arguments we preceed them with
+=&optional= marker just as we do in the argument list.
+
+#+BEGIN_SRC elisp
+(defun drop-items (list &optional n)
+  "Drop first item of LIST or N items if N is provided."
+  (declare (elsa ((list a*) &optional int) (list a*)))
+  (setq n (or n 1))
+  (dotimes (_ n list)
+    (setq list (cdr list))))
+#+END_SRC
+
+** Variadic types
+
+If a function can take arbitrary number of arguments we preceed the
+last variadic argument with =&rest= marker just as we do in the argument
+list.
+
+#+BEGIN_SRC elisp
+(defun join (separator &rest strings)
+  "Join STRINGS with SEPARATOR."
+  (declare (elsa (string &rest string) string))
+  (mapconcat 'identity strings separator))
+#+END_SRC
+
+** Type combinators
+*** Sum types
+:PROPERTIES:
+:ID:       5a21a68a-4df1-4d44-a854-1d9700858a1a
+:END:
+
+Sum types can be specified as a list form starting with =or=, so =(or
+string int)= is a type accepting strings or integers.
+
+A sum type is useful if the function internally checks the passed
+value and decides what processing to do:
+
+*TODO:* I think the double parens in the "argument" portion will lead to
```

Comment:
I guess this is the same problem as the double-parenthesis in the bindings of the `let` form. I don't find it confusing but I don't know about others.

Reply:
More than confusing it's also a bit noisy I think.

    (elsa ((function (a*) b*) (repeat a*)) (repeat b*))

for me it's very hard to get a good overview of where the arguments are and where the return type is.

    (elsa (function (a*) b*) (repeat a*) -> (repeat b*))

is to me a lot more visual.

In my original thoughts I actually had two annotations `(elsa-args)` and `(elsa-return)`.  Not sure if that's better though.

Reply:
It seems to me the idiomatic lisp thing would be to break the s-expression onto multiple lines and use indentation to visually indicate structure. The `->` notation is nice and efficient for one-line type annotations, but in the real world it seems likely that people would get that visual reinforcement of the type structure by writing 
```elisp
(elsa ((function (a*) b*)
       (repeat a*))
    (repeat b*))
```
instead of
```elisp
(elsa ((function (a*) b*) (repeat a*)) (repeat b*))
```
at least within elisp code.

Reply:
This is really nice idea and should be easily implemented using the indent engine.  When you lay out out like this the structure actually makes sense too!

I will update the examples to use this layout and hopefully it catches on.